### DISCUSSION TITLE
OIDC OAuth2 Setup with Google (also, debug logging?)

### DISCUSSION THREAD

**Discussion by APTy** [lowdefy/lowdefy#687] [Questions and Help] [answered]
This is a two-parter:

1. Is there a way to turn on more detailed debug logging on the server? Sometimes when I'm debugging an issue it'd be nice to get more detail on what failed.

2. This is mostly related to setting up `oidc` with [Google](https://developers.google.com/identity/protocols/oauth2/openid-connect) and having trouble understanding what is going on under the hood. My current error:
```json
{
  "errors": [
    {
      "message": "Issuer.discover() failed.\n    s: expected 200 OK, got: 404 Not Found\n    s: expected 200 OK, got: 404 Not Found",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "openIdAuthorizationUrl"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "name": "AggregateError",
          "stacktrace": [
            "AggregateError: Issuer.discover() failed.",
            "    s: expected 200 OK, got: 404 Not Found",
            "    s: expected 200 OK, got: 404 Not Found",
            "    at Function.discover (/path/to/.lowdefy/.cache/scripts/graphql/3_18_0/package/dist/240.index.js:28:3738643)",
            "    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
            "    at async _.getClient (/path/to/.lowdefy/.cache/scripts/graphql/3_18_0/package/dist/253.index.js:84:4269)",
            "    at async _.getAuthorizationUrl (/path/to/.lowdefy/.cache/scripts/graphql/3_18_0/package/dist/253.index.js:84:4792)"
          ]
        }
      }
    }
  ],
  "data": {
    "openIdAuthorizationUrl": null
  }
}
```

my env:
```bash
export LOWDEFY_SECRET_OPENID_CLIENT_ID=<my_client_id>.apps.googleusercontent.com
export LOWDEFY_SECRET_OPENID_CLIENT_SECRET=<my_secret>
export LOWDEFY_SECRET_OPENID_DOMAIN=https://accounts.google.com/o/oauth2/v2/auth
```

It _seems_ like lowdefy is failing to find the openid issuer, but I don't really have any intuition on why, since the `LOWDEFY_SECRET_OPENID_DOMAIN` definitely isn't 404ing. So I'd be curious which URL is actually 404ing, for example.

**Comment by Gervwyk** [comment, marked as answer]
Hi @APTy 
On point 2. I would change the `.env` file to:
```
LOWDEFY_SECRET_OPENID_CLIENT_ID="<my_client_id>.apps.googleusercontent.com"
LOWDEFY_SECRET_OPENID_CLIENT_SECRET="<my_secret>"
LOWDEFY_SECRET_OPENID_DOMAIN="https://accounts.google.com/.well-known/openid-configuration"
LOWDEFY_SECRET_JWT_SECRET="<my_jwt_secret>"
```
I suspect that your `LOWDEFY_SECRET_OPENID_DOMAIN` is incorrect. Please let me know if this resolves it?

On point 1, yes, a more verbose server logging mode will be very useful. I've added this as a v4 feature #688

**Reply by APTy** [reply]
@Gervwyk thank you! I had a feeling I was using the wrong URL! Is there any way we can update the docs to say something like "the LOWDEFY_SECRET_OPENID_DOMAIN var should point to the openid configuration URL, such as `.../.well-known/openid-configuration`"

**Reply by Gervwyk** [reply]
That's good feedback, thanks! I'll check with @SamTolmay he has been through a few of these setups. Hopefully real soon we can add provider-specific "how-to" guides to make it easier as well for new dev. Glad it worked 💫

**Reply by krg-digital** [reply]
@Gervwyk Was wondering where you pulled that OpenID url for Google from? I'm trying to set up a Google sign-in option and can't really find that kind of info in the Google documentation. Likewise, do you happen to know where I can find the logout redirect uri for Google?

**Reply by SamTolmay** [reply]
Hi @krg-digital 

The documentation on the Google OpenID Connect discovery URL can be found here: https://developers.google.com/identity/protocols/oauth2/openid-connect#discovery

When using Google as your OpenID Connect provider you generally wouldn't want to log out the user from Google (I don't think Google provides a URL to do this but might be wrong).

If you don't set a logout URL, the user will be redirected to your app's home page after they log out. You could also set any other URL as the logout URL, like your website or a "login" page in your app.

So, for example:
```yaml
config:
  auth:
    openId:
      logoutRedirectUri: /login
```
or
```yaml
config:
  auth:
    openId:
      logoutRedirectUri: https://www.example.com
```

Hope this helps 😀
